README.md
# yii2-domain-validator

Domain validator for Yii 2.

[![License](https://poser.pugx.org/kdn/yii2-domain-validator/license)](https://packagist.org/packages/kdn/yii2-domain-validator)
[![Latest Stable Version](https://poser.pugx.org/kdn/yii2-domain-validator/v/stable)](https://packagist.org/packages/kdn/yii2-domain-validator)
[![Code Coverage](https://scrutinizer-ci.com/g/dmitry-kulikov/yii2-domain-validator/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/dmitry-kulikov/yii2-domain-validator/?branch=master)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/dmitry-kulikov/yii2-domain-validator/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/dmitry-kulikov/yii2-domain-validator/?branch=master)
[![Code Climate](https://codeclimate.com/github/dmitry-kulikov/yii2-domain-validator/badges/gpa.svg)](https://codeclimate.com/github/dmitry-kulikov/yii2-domain-validator)

## Requirements

- PHP 5.4 or later or HHVM 3;
- Yii framework 2;
- PHP extensions:
  - `ctype` (character type checking) extension (required);
  - `mbstring` (multibyte string) extension (required);
  - `intl` (internationalization functions) extension (optional, for IDN only).

## Installation

The preferred way to install this extension is through [Composer](https://getcomposer.org).

To install, either run

```sh
php composer.phar require kdn/yii2-domain-validator "*"
```

or add

```text
"kdn/yii2-domain-validator": "*"
```

to the `require` section of your `composer.json` file.

## Usage

Model class example:

```php
<?php

namespace app\models;

use kdn\yii2\validators\DomainValidator;
use Yii;
use yii\base\Model;

class YourCustomModel extends Model
{
    public $domain;

    public function rules()
    {
        return [
            ['domain', DomainValidator::class],
            /*
            or with custom options: enable IDN and forbid URLs
            [
                'domain',
                DomainValidator::class,
                'enableIDN' => true,
                'allowURL' => false,
            ],
            */
        ];
    }

    public function attributeLabels()
    {
        return [
            'domain' => Yii::t('app', 'Domain Name'),
        ];
    }
}
```

Please view public properties in class
[DomainValidator](https://github.com/dmitry-kulikov/yii2-domain-validator/blob/master/src/DomainValidator.php)
to get info about all available options, they documented comprehensively. Here I will highlight only non-evident things.

1. By default, validator allows URL, it will try to parse URL and then validate domain name.
   Note that model attribute value itself will not be modified.
   If URL parsing fails then validator considers value as domain.
   Validator may work not perfect for invalid URLs. For example user input is `http//example.com`,
   the error message will be `Each label of the input value can consist of only letters, numbers and hyphens`,
   although it would be better to show something like `Invalid URL`.
   The problem is that if field allows both URL and bare domain name and the input value is invalid,
   then it is impossible to reliably determine what did user want `http://example.com` or `http.example.com`.
   If you don't need URLs at all, only stand-alone domain name, you can disable this behavior
   by setting `allowURL` to `false`.
   If you always need to validate domain name in URL, no stand-alone domain name,
   then you should add URL validator before domain name validator:

   ```php
   public function rules()
   {
       return [
           ['domain', 'url'],
           ['domain', DomainValidator::class],
       ];
   }
   ```

1. By default, minimum number of domain name labels is 2. So `example` - invalid, `example.com` - valid.
   It is not standard requirement for domain name, standard states that domain name `example` is valid.
   I added this restriction for practical reasons, you can disable it or require even more domain name labels
   using option `labelNumberMin`.
1. Client side validation not implemented, and I have not such plans.
   Please consider [AJAX validation](https://www.yiiframework.com/doc/guide/2.0/en/input-validation#ajax-validation)
   if you want to bring domain validation on client side.

## Testing

Make sure you installed all Composer dependencies (run `composer update` in the base directory of repository).
Run PHPUnit in the base directory of repository:

```sh
./vendor/bin/phpunit
```

### Testing using Docker

#### Requirements

- Docker >= 19.03.0 ([install](https://docs.docker.com/get-docker));
- Docker Compose >= 1.25.5 ([install](https://docs.docker.com/compose/install));
- Docker plugins:
  - buildx ([install](https://github.com/docker/buildx#installing)).

#### Up and running

1. Provide credentials for Composer:

   ```sh
   cp auth.json.example \
       auth.json
   ```

   I suggest to set GitHub OAuth token (also known as personal access token) in `auth.json`,
   however if you have doubts about security, or you are lazy to generate token then you can replace content of
   `auth.json` on `{}`, in most cases this will work.

1. Build images for services:

   ```sh
   docker buildx bake --load --pull
   ```

   or

   ```sh
   docker buildx bake --load --pull --no-cache --progress plain
   ```

   see `docker buildx bake --help` for details.

1. Start service in background mode:

   ```sh
   docker-compose up --detach 8.1
   ```

   This command will start the service with PHP 8.1. Also allowed `7.4`, `5.6`, `8.1-alpine`, `7.4-alpine`
   and `5.6-alpine`, see services defined in `docker-compose.yml`.

1. Execute tests in the running container:

   ```sh
   docker-compose exec 8.1 ./vendor/bin/phpunit
   ```

   Alternatively you can start a shell in the running container and execute tests from it:

   ```sh
   docker-compose exec 8.1 sh
   $ ./vendor/bin/phpunit
   ```

1. Stop and remove containers created by `up`:

   ```sh
   docker-compose down
   ```

   You may want to remove volumes along with containers:

   ```sh
   docker-compose down --volumes
   ```

## Backward compatibility promise

yii2-domain-validator is using [Semver](https://semver.org). This means that versions are tagged
with MAJOR.MINOR.PATCH. Only a new major version will be allowed to break backward
compatibility (BC).

PHP 8 introduced [named arguments](https://wiki.php.net/rfc/named_params), which
increased the cost and reduces flexibility for package maintainers. The names of the
arguments for methods in yii2-domain-validator is not included in our BC promise.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.4] - 2022-06-11

### Changed

- Running of tests using Docker and documentation for this.

## [1.1.3] - 2021-10-22

### Changed

- Running of tests using Docker and documentation for this.

## [1.1.2] - 2020-02-19

### Added

- Added possibility to run tests using Docker and documentation for this.

## [1.1.1] - 2019-11-02

### Fixed

- checkDNS = true leads to ErrorException for some bad domains on Alpine Linux.

## [1.1.0] - 2019-06-11

### Added

- \#5: Added the ability to extend the verification of DNS (maranqz).

### Fixed

- Fixed `kdn\yii2\validators\DomainValidator::$checkDNS` tells that every domain is correct on Alpine Linux.

## [1.0.3] - 2018-03-07

### Fixed

- \#4: Fixed error for $allowUnderscore = true in PHP 7.3.

## [1.0.2] - 2017-02-19

### Fixed

- \#3: Fixed checkDNS returns false positives.
- Fixed typo in Russian translation.

## [1.0.1] - 2016-07-15

### Added

- Documentation.

## [1.0.0] - 2016-07-07

- Initial release.


run_tests.sh
#!/usr/bin/env sh

PHP_VERSION=$1
NAME="yii2-domain-validator:php-${PHP_VERSION}"
sudo docker build --build-arg PHP_VERSION="${PHP_VERSION}" --tag "${NAME}" .
sudo docker run --interactive --rm --tty "${NAME}"
sudo docker image prune --force


tests/TestCase.php
<?php

namespace kdn\yii2\validators;

use PHPUnit_Framework_TestCase;
use Yii;
use yii\helpers\ArrayHelper;

/**
 * Class TestCase.
 * @package kdn\yii2\validators
 */
abstract class TestCase extends PHPUnit_Framework_TestCase
{
    /**
     * Mock Yii application.
     * @before
     */
    protected function prepare()
    {
        static::mockApplication();
    }

    /**
     * Clean up after test.
     * By default, the application created with `mockApplication` will be destroyed.
     * @after
     */
    protected function clear()
    {
        static::destroyApplication();
    }

    /**
     * Populates Yii::$app with a new application.
     * The application will be destroyed on clear() automatically.
     * @param array $config the application configuration, if needed
     * @param string $appClass name of the application class to create
     */
    protected static function mockApplication($config = [], $appClass = 'yii\console\Application')
    {
        new $appClass(
            ArrayHelper::merge(
                [
                    'id' => 'test-app',
                    'basePath' => __DIR__,
                    'vendorPath' => static::getVendorPath(),
                ],
                $config
            )
        );
    }

    /**
     * Get path to "vendor" directory.
     * @return string path to "vendor" directory.
     */
    protected static function getVendorPath()
    {
        return dirname(__DIR__) . '/vendor';
    }

    /**
     * Destroys application in Yii::$app by setting it to null.
     */
    protected static function destroyApplication()
    {
        Yii::$app = null;
    }

    /**
     * Add column to array.
     * @param array $array
     * @param mixed $value
     * @return array
     */
    protected static function arrayAddColumn($array, $value)
    {
        return array_map(
            function ($data) use ($value) {
                $data[] = $value;

                return $data;
            },
            $array
        );
    }

    /**
     * \u escape sequence for PHP.
     * @param string $text
     * @return string
     */
    protected static function u($text)
    {
        return json_decode("\"$text\"");
    }
}


tests/bootstrap.php
<?php

// ensure we get report on all possible php errors
error_reporting(-1);

require_once(__DIR__ . '/../vendor/autoload.php');
require_once(__DIR__ . '/../vendor/yiisoft/yii2/Yii.php');


messages/ru/domain.php
<?php

/**
 * Message translations.
 *
 * Each array element represents the translation (value) of a message (key).
 * If the value is empty, the message is considered as not translated.
 * Messages that no longer need translation will have their translations
 * enclosed between a pair of '@@' marks.
 *
 * Message string can be used with plural forms format. Check i18n section
 * of the Yii 2 guide for details.
 *
 * NOTE: this file must be saved in UTF-8 encoding.
 */

return [
    'Domain' => 'Домен',
    'Domain Name' => 'Доменное имя',
    '{attribute} is invalid.' => 'Значение «{attribute}» неверно.',
    'DNS record corresponding to {attribute} not found.' => 'DNS запись, соответствующая значению «{attribute}», не найдена.',
    '{attribute} should contain at least 1 character.' => 'Значение «{attribute}» должно содержать минимум 1 символ.',
    '{attribute} should consist of at least {labelNumberMin, number} labels separated by {labelNumberMin, plural, =2{dot} other{dots}}.' => 'Значение «{attribute}» должно состоять минимум из {labelNumberMin, number} {labelNumberMin, plural, one{метки} other{меток}}, разделённых {labelNumberMin, plural, =2{точкой} other{точками}}.',
    'Each label of {attribute} should contain at least 1 character.' => 'Каждая метка значения «{attribute}» должна содержать минимум 1 символ.',
    '{attribute} must be a string.' => 'Значение «{attribute}» должно быть строкой.',
    'Each label of {attribute} should start and end with letter or number. The rightmost label of {attribute} should start with letter.' => 'Каждая метка значения «{attribute}» должна начинаться и заканчиваться буквой или цифрой. Крайняя правая метка значения «{attribute}» должна начинаться с буквы.',
    'Each label of {attribute} should start and end with latin letter or number. The rightmost label of {attribute} should start with latin letter.' => 'Каждая метка значения «{attribute}» должна начинаться и заканчиваться латинской буквой или цифрой. Крайняя правая метка значения «{attribute}» должна начинаться с латинской буквы.',
    'Label of {attribute} is too long.' => 'Метка значения «{attribute}» слишком длинна.',
    'Each label of {attribute} should contain at most 63 characters.' => 'Каждая метка значения «{attribute}» должна содержать максимум 63 символа.',
    '{attribute} is too long.' => 'Значение «{attribute}» слишком длинно.',
    '{attribute} should contain at most 253 characters.' => 'Значение «{attribute}» должно содержать максимум 253 символа.',
    'Each label of {attribute} can consist of only letters, numbers, hyphens and underscores.' => 'Каждая метка значения «{attribute}» может состоять только из букв, цифр, дефисов и символов подчёркивания.',
    'Each label of {attribute} can consist of only latin letters, numbers, hyphens and underscores.' => 'Каждая метка значения «{attribute}» может состоять только из латинских букв, цифр, дефисов и символов подчёркивания.',
    'Each label of {attribute} can consist of only letters, numbers and hyphens.' => 'Каждая метка значения «{attribute}» может состоять только из букв, цифр и дефисов.',
    'Each label of {attribute} can consist of only latin letters, numbers and hyphens.' => 'Каждая метка значения «{attribute}» может состоять только из латинских букв, цифр и дефисов.',
];


src/DomainValidator.php
<?php

namespace kdn\yii2\validators;

use Yii;
use yii\base\ErrorException;
use yii\base\InvalidConfigException;
use yii\validators\Validator;

/**
 * Class DomainValidator.
 * @package kdn\yii2\validators
 */
class DomainValidator extends Validator
{
    /**
     * @var bool whether to allow underscores in domain name;
     * defaults to false
     */
    public $allowUnderscore = false;

    /**
     * @var bool whether to allow the URL address along with domain name;
     * defaults to true, meaning that validator should try to parse URL address and then validate domain name
     */
    public $allowURL = true;

    /**
     * @var bool|callable whether to check whether domain name exists;
     * be aware that this check can fail due to temporary DNS problems even if domain name exists;
     * do not use it to check domain name availability;
     * defaults to false;
     * this field can be specified as a PHP callback, for example:
     * ```php
     * function (string $domain) {
     *     $records = @dns_get_record("$domain.", DNS_MX); // @ is just for simplicity of example, avoid using it
     *     if (empty($records)) {
     *         return ['Cannot find Mail Exchanger record for "{value}".', ['value' => $domain]];
     *     }
     *
     *     return null; // the data is valid
     * }
     * ```
     * note that alternatively you can override method `checkDNS`
     * @see checkDNS
     */
    public $checkDNS = false;

    /**
     * @var bool whether validation process should take into account IDN (internationalized domain names);
     * defaults to false, meaning that validation of domain names containing IDN will always fail;
     * note that in order to use IDN validation you have to install and enable `intl` PHP extension,
     * otherwise an exception would be thrown
     */
    public $enableIDN = false;

    /**
     * @var string the encoding of the string value to be validated (e.g. 'UTF-8');
     * if this property is not set, `\yii\base\Application::charset` will be used
     */
    public $encoding;

    /**
     * @var string the base path for all translated messages; specify it if you want to use custom translated messages
     */
    public $i18nBasePath;

    /**
     * @var int minimum number of domain name labels;
     * defaults to 2, meaning that domain name should contain at least 2 labels
     * @see messageLabelNumberMin for the customized message for domain name with too small number of labels
     */
    public $labelNumberMin = 2;

    /**
     * @var string user-defined error message used when domain name is invalid but
     * reason is too complicated for explanation to end-user or details are not needed at all;
     * you may use the following placeholders in the message:
     * - `{attribute}`: the label of the attribute being validated
     * - `{value}`: the value of the attribute being validated
     * @see simpleErrorMessage to use this message for all possible errors
     */
    public $message;

    /**
     * @var string user-defined error message used when DNS record corresponding to domain name not found;
     * you may use the following placeholders in the message:
     * - `{attribute}`: the label of the attribute being validated
     * - `{value}`: the value of the attribute being validated
     */
    public $messageDNS;

    /**
     * @var string user-defined error message used when domain name contains an invalid character;
     * you may use the following placeholders in the message:
     * - `{attribute}`: the label of the attribute being validated
     * - `{value}`: the value of the attribute being validated
     */
    public $messageInvalidCharacter;

    /**
     * @var string user-defined error message used when number of domain name labels is smaller than `labelNumberMin`;
     * you may use the following placeholders in the message:
     * - `{attribute}`: the label of the attribute being validated
     * - `{labelNumberMin}`: the value of `labelNumberMin`
     * - `{value}`: the value of the attribute being validated
     */
    public $messageLabelNumberMin;

    /**
     * @var string user-defined error message used when domain name label starts or ends with an invalid character;
     * you may use the following placeholders in the message:
     * - `{attribute}`: the label of the attribute being validated
     * - `{value}`: the value of the attribute being validated
     */
    public $messageLabelStartEnd;

    /**
     * @var string user-defined error message used when domain name label is too long;
     * you may use the following placeholders in the message:
     * - `{attribute}`: the label of the attribute being validated
     * - `{value}`: the value of the attribute being validated
     */
    public $messageLabelTooLong;

    /**
     * @var string user-defined error message used when domain name label is too short;
     * you may use the following placeholders in the message:
     * - `{attribute}`: the label of the attribute being validated
     * - `{value}`: the value of the attribute being validated
     */
    public $messageLabelTooShort;

    /**
     * @var string user-defined error message used when domain name is not a string;
     * you may use the following placeholders in the message:
     * - `{attribute}`: the label of the attribute being validated
     * - `{value}`: the value of the attribute being validated
     */
    public $messageNotString;

    /**
     * @var string user-defined error message used when domain name is too long;
     * you may use the following placeholders in the message:
     * - `{attribute}`: the label of the attribute being validated
     * - `{value}`: the value of the attribute being validated
     */
    public $messageTooLong;

    /**
     * @var string user-defined error message used when domain name is too short;
     * you may use the following placeholders in the message:
     * - `{attribute}`: the label of the attribute being validated
     * - `{value}`: the value of the attribute being validated
     */
    public $messageTooShort;

    /**
     * @var bool whether to always use simple error message;
     * defaults to false, meaning that validator should use specialized error messages for different errors,
     * it should help end-user to understand reason of error; set it to true if detailed error messages don't fit
     * for your application then `message` will be used in all cases
     */
    public $simpleErrorMessage = false;

    /**
     * {@inheritdoc}
     */
    public function init()
    {
        parent::init();
        if ($this->enableIDN && !function_exists('idn_to_ascii')) {
            throw new InvalidConfigException(
                'In order to use IDN validation intl extension must be installed and enabled.'
            );
        }
        if (!isset($this->encoding)) {
            $this->encoding = Yii::$app->charset;
        }
        if (!isset($this->i18nBasePath)) {
            $this->i18nBasePath = dirname(__DIR__) . '/messages';
        }
        Yii::$app->i18n->translations['kdn/yii2/validators/domain'] = [
            'class' => 'yii\i18n\PhpMessageSource',
            'basePath' => $this->i18nBasePath,
            'fileMap' => ['kdn/yii2/validators/domain' => 'domain.php'],
        ];
    }

    /**
     * {@inheritdoc}
     */
    protected function validateValue($value)
    {
        if (!is_string($value)) {
            return $this->getErrorMessage('messageNotString');
        }

        if (empty($value)) {
            return $this->getErrorMessage('messageTooShort');
        }

        if ($this->allowURL) {
            $host = parse_url($value, PHP_URL_HOST);
            if (isset($host) && $host !== false) {
                $value = $host;
            }
        }

        if ($this->enableIDN) {
            $idnaInfo = [];
            $options = IDNA_CHECK_BIDI | IDNA_CHECK_CONTEXTJ;
            $asciiValue = idn_to_ascii($value, $options, INTL_IDNA_VARIANT_UTS46, $idnaInfo);
            if ($asciiValue !== false) {
                $value = $asciiValue;
            } else {
                $idnaErrors = null;
                if (is_array($idnaInfo) && array_key_exists('errors', $idnaInfo)) {
                    $idnaErrors = $idnaInfo['errors'];
                }
                if ($idnaErrors & IDNA_ERROR_DOMAIN_NAME_TOO_LONG) {
                    $errorMessageName = 'messageTooLong';
                } elseif ($idnaErrors & IDNA_ERROR_EMPTY_LABEL) {
                    $errorMessageName = 'messageLabelTooShort';
                } elseif ($idnaErrors & IDNA_ERROR_LABEL_TOO_LONG) {
                    $errorMessageName = 'messageLabelTooLong';
                } elseif ($idnaErrors & IDNA_ERROR_DISALLOWED) {
                    $errorMessageName = 'messageInvalidCharacter';
                } elseif ($idnaErrors & IDNA_ERROR_LEADING_HYPHEN || $idnaErrors & IDNA_ERROR_TRAILING_HYPHEN) {
                    $errorMessageName = 'messageLabelStartEnd';
                } elseif (empty($idnaInfo)) {
                    // too long domain name caused buffer overflow
                    $errorMessageName = 'messageTooLong';
                } else {
                    $errorMessageName = 'message';
                }

                return $this->getErrorMessage($errorMessageName);
            }
        }

        // ignore trailing dot
        if (mb_substr($value, -1, 1, $this->encoding) == '.') {
            $value = substr_replace($value, '', -1);
        }

        // 253 characters limit is same as 127 levels,
        // domain name with 127 levels with 1 character per label will be 253 characters long
        if (mb_strlen($value, $this->encoding) > 253) {
            return $this->getErrorMessage('messageTooLong');
        }

        $labels = explode('.', $value);
        $labelsCount = count($labels);

        if ($labelsCount < $this->labelNumberMin) {
            return $this->getErrorMessage('messageLabelNumberMin', ['labelNumberMin' => $this->labelNumberMin]);
        }

        for ($i = 0; $i < $labelsCount; $i++) {
            $label = $labels[$i];
            $labelLength = mb_strlen($label, $this->encoding);

            if (empty($label)) {
                return $this->getErrorMessage('messageLabelTooShort');
            }

            if ($labelLength > 63) {
                return $this->getErrorMessage('messageLabelTooLong');
            }

            if ($this->allowUnderscore) {
                $pattern = '/^[a-z\d\-_]+$/i';
            } else {
                $pattern = '/^[a-z\d\-]+$/i';
            }
            if (!preg_match($pattern, $label)) {
                return $this->getErrorMessage('messageInvalidCharacter');
            }

            if (
                $i == $labelsCount - 1 && !ctype_alpha($label[0])
                || !ctype_alnum($label[0])
                || !ctype_alnum($label[$labelLength - 1])
            ) {
                return $this->getErrorMessage('messageLabelStartEnd');
            }
        }

        if ($this->checkDNS) {
            if (is_callable($this->checkDNS)) {
                return call_user_func($this->checkDNS, $value);
            }

            if (!$this->checkDNS($value)) {
                return $this->getErrorMessage('messageDNS');
            }
        }

        return null;
    }

    /**
     * Check whether domain name exists.
     * @param string $domain domain name
     * @return bool whether domain name exists.
     * @see https://github.com/yiisoft/yii2/issues/17083
     * @see https://github.com/yiisoft/yii2/issues/17602
     */
    protected function checkDNS($domain)
    {
        $normalizedDomain = "$domain.";
        if (!checkdnsrr($normalizedDomain, 'ANY')) {
            return false;
        }

        try {
            // dns_get_record can return false and emit Warning that may or may not be converted to ErrorException
            $records = dns_get_record($normalizedDomain, DNS_ANY);
        } catch (ErrorException $e) {
            return false;
        }

        return !empty($records);
    }

    /**
     * Get error message by name.
     * @param string $name error message name
     * @param array $params parameters to be inserted into the error message
     * @return array error message.
     */
    protected function getErrorMessage($name, $params = [])
    {
        if ($this->simpleErrorMessage) {
            $name = 'message';
        }
        if (isset($this->$name)) {
            return [$this->$name, $params];
        }
        $this->$name = Yii::t('kdn/yii2/validators/domain', $this->getDefaultErrorMessages()[$name]);

        return [$this->$name, $params];
    }

    /**
     * Get default error messages.
     * @return array default error messages.
     */
    protected function getDefaultErrorMessages()
    {
        $messages = [
            'message' => '{attribute} is invalid.',
            'messageDNS' => 'DNS record corresponding to {attribute} not found.',
            'messageLabelNumberMin' =>
                '{attribute} should consist of at least {labelNumberMin, number} labels separated by ' .
                '{labelNumberMin, plural, =2{dot} other{dots}}.',
            'messageLabelTooShort' => 'Each label of {attribute} should contain at least 1 character.',
            'messageNotString' => '{attribute} must be a string.',
            'messageTooShort' => '{attribute} should contain at least 1 character.',
        ];
        if ($this->enableIDN) {
            $messages['messageLabelStartEnd'] =
                'Each label of {attribute} should start and end with letter or number.' .
                ' The rightmost label of {attribute} should start with letter.';
            $messages['messageLabelTooLong'] = 'Label of {attribute} is too long.';
            $messages['messageTooLong'] = '{attribute} is too long.';
            if ($this->allowUnderscore) {
                $messages['messageInvalidCharacter'] =
                    'Each label of {attribute} can consist of only letters, numbers, hyphens and underscores.';
            } else {
                $messages['messageInvalidCharacter'] =
                    'Each label of {attribute} can consist of only letters, numbers and hyphens.';
            }
        } else {
            $messages['messageLabelStartEnd'] =
                'Each label of {attribute} should start and end with latin letter or number.' .
                ' The rightmost label of {attribute} should start with latin letter.';
            $messages['messageLabelTooLong'] = 'Each label of {attribute} should contain at most 63 characters.';
            $messages['messageTooLong'] = '{attribute} should contain at most 253 characters.';
            if ($this->allowUnderscore) {
                $messages['messageInvalidCharacter'] =
                    'Each label of {attribute} can consist of only latin letters, numbers, hyphens and underscores.';
            } else {
                $messages['messageInvalidCharacter'] =
                    'Each label of {attribute} can consist of only latin letters, numbers and hyphens.';
            }
        }

        return $messages;
    }
}
